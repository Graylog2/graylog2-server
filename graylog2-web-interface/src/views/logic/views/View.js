// @flow strict
import * as Immutable from 'immutable';
import ObjectID from 'bson-objectid';

import ViewState from './ViewState';
import Search from '../search/Search';
import type { QueryId } from '../queries/Query';

export type Properties = Immutable.List<any>;

export type PluginMetadata = {
  name: string,
  url: string,
};
export type Requirements = { [string]: PluginMetadata };
export type ViewStateMap = Immutable.Map<QueryId, ViewState>;

export type SearchType = 'SEARCH';
export type DashboardType = 'DASHBOARD';
export type ViewType = SearchType | DashboardType;

type InternalState = {
  id: string,
  type: ViewType,
  title: string,
  summary: string,
  description: string,
  search: Search,
  properties: Properties,
  state: ViewStateMap,
  createdAt: Date,
  owner: string,
  requires: Requirements,
};

export type WidgetMapping = Immutable.Map<string, string>;
export type ViewJson = {
  id: string,
  type: ViewType,
  title: string,
  summary: string,
  description: string,
  search_id: string,
  properties: Properties,
  state: { [QueryId]: ViewState },
  created_at: Date,
  owner: string,
  requires: Requirements,
};

export default class View {
  static Type: { Search: SearchType, Dashboard: DashboardType } = {
    Search: 'SEARCH',
    Dashboard: 'DASHBOARD',
  };

  _value: InternalState;

  constructor(id: string,
    type: ViewType,
    title: string,
    summary: string,
    description: string,
    search: Search,
    properties: Properties,
    state: ViewStateMap,
    createdAt: Date,
    owner: string,
    requires: Requirements) {
    this._value = {
      id,
      type,
      title,
      summary,
      description,
      search,
      properties: Immutable.fromJS(properties),
      state: Immutable.fromJS(state),
      createdAt,
      owner,
      requires,
    };
  }

  static create(): View {
    // eslint-disable-next-line no-use-before-define
    return new Builder().createdAt(new Date()).build();
  }

  get id(): string {
    return this._value.id;
  }

  get type(): ViewType {
    return this._value.type;
  }

  get title(): string {
    return this._value.title;
  }

  get summary(): string {
    return this._value.summary;
  }

  get description(): string {
    return this._value.description;
  }

  get search(): Search {
    return this._value.search;
  }

  get properties(): Properties {
    return this._value.properties;
  }

  get state(): ViewStateMap {
    return this._value.state;
  }

  get createdAt(): Date {
    return this._value.createdAt;
  }

  get widgetMapping(): WidgetMapping {
    return (this.state || Immutable.Map()).valueSeq().map(s => s.widgetMapping).reduce((prev, cur) => Immutable.fromJS(prev).merge(Immutable.fromJS(cur)));
  }

  get owner(): string {
    return this._value.owner;
  }

  get requires(): Requirements {
    return this._value.requires || {};
  }

  // eslint-disable-next-line no-use-before-define
  toBuilder(): Builder {
    const { id, title, summary, description, search, properties, state, createdAt, owner, requires, type } = this._value;
    // eslint-disable-next-line no-use-before-define
    return new Builder(Immutable.Map({
      id,
      title,
      summary,
      description,
      search,
      properties,
      state,
      createdAt,
      owner,
      requires,
      type,
    }));
  }

  toJSON() {
    const { id, type, title, summary, description, search, properties, state, createdAt, owner } = this._value;

    return {
      id,
      type,
      title,
      summary,
      description,
      search_id: search.id,
      properties,
      state,
      created_at: createdAt,
      owner,
    };
  }

  static fromJSON(value: ViewJson): View {
    // eslint-disable-next-line camelcase
    const { id, type, title, summary, description, properties, state, created_at, owner, requires } = value;
    const viewState: ViewStateMap = Immutable.Map(state).map(ViewState.fromJSON);
    return View.create()
      .toBuilder()
      .id(id)
      .type(type)
      .title(title)
      .summary(summary)
      .description(description)
      .properties(properties)
      .state(viewState)
      .createdAt(created_at)
      .owner(owner)
      .requires(requires)
      .build();
  }

  // eslint-disable-next-line no-use-before-define
  static builder(): Builder {
    // eslint-disable-next-line no-use-before-define
    return new Builder();
  }
}

class Builder {
  value: Immutable.Map<string, any>;

  constructor(value: Immutable.Map = Immutable.Map()) {
    this.value = value;
  }

  id(value: string): Builder {
    return new Builder(this.value.set('id', value));
  }

  toNewView(): Builder {
    return new Builder(this.value.set('id', undefined).set('title', undefined));
  }

  type(value: ViewType): Builder {
    return new Builder(this.value.set('type', value));
  }

  newId(): Builder {
    return this.id(ObjectID().toString());
  }

  title(value: string): Builder {
    return new Builder(this.value.set('title', value));
  }

  summary(value: string): Builder {
    return new Builder(this.value.set('summary', value));
  }

  description(value: string): Builder {
    return new Builder(this.value.set('description', value));
  }

  search(value: Search): Builder {
    return new Builder(this.value.set('search', value));
  }

  properties(value: Properties): Builder {
    return new Builder(this.value.set('properties', value));
  }

  state(value: ViewStateMap): Builder {
    return new Builder(this.value.set('state', Immutable.fromJS(value)));
  }

  createdAt(value: Date): Builder {
    return new Builder(this.value.set('createdAt', value));
  }

  owner(value: string): Builder {
    return new Builder(this.value.set('owner', value));
  }

  requires(value: Requirements): Builder {
    return new Builder(this.value.set('requires', value));
  }

  build(): View {
    const { id, type, title, summary, description, search, properties, state, createdAt, owner, requires } = this.value.toObject();
    return new View(id, type, title, summary, description, search, properties, state, createdAt, owner, requires);
  }
}
