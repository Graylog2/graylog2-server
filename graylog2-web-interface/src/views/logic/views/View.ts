/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as Immutable from 'immutable';
import { flatten } from 'lodash';
import ObjectID from 'bson-objectid';

import Widget from 'views/logic/widgets/Widget';
import defaultTitle from 'views/components/defaultTitle';

import ViewState from './ViewState';
import type { WidgetMapping } from './types';
import type { ViewStateJson } from './ViewState';

import Search from '../search/Search';
import type { SearchType as QuerySearchType } from '../queries/SearchType';
import type { QueryId } from '../queries/Query';

export type Properties = Immutable.List<any>;

export type PluginMetadata = {
  name: string,
  url: string,
};
export type Requirements = { [key: string]: PluginMetadata };
export type ViewStateMap = Immutable.Map<QueryId, ViewState>;

export type SearchType = 'SEARCH';
export type DashboardType = 'DASHBOARD';
export type ViewType = SearchType | DashboardType;

type InternalState = {
  id: string,
  type: ViewType,
  title: string,
  summary: string,
  description: string,
  search: Search,
  properties: Properties,
  state: ViewStateMap,
  createdAt: Date,
  owner: string,
  requires: Requirements,
};

export type ViewJson = {
  id: string,
  type: ViewType,
  title: string,
  summary: string,
  description: string,
  search_id: string,
  properties: Properties,
  state: { [key: string]: ViewStateJson },
  created_at: Date,
  owner: string,
  requires: Requirements,
};

export default class View {
  static Type: { Search: SearchType, Dashboard: DashboardType } = {
    Search: 'SEARCH',
    Dashboard: 'DASHBOARD',
  };

  _value: InternalState;

  constructor(id: string,
    type: ViewType,
    title: string,
    summary: string,
    description: string,
    search: Search,
    properties: Properties,
    state: ViewStateMap,
    createdAt: Date,
    owner: string,
    requires: Requirements) {
    this._value = {
      id,
      type,
      title,
      summary,
      description,
      search,
      properties: Immutable.fromJS(properties),
      state: Immutable.fromJS(state),
      createdAt,
      owner,
      requires,
    };
  }

  static create(): View {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder().createdAt(new Date()).build();
  }

  get id(): string {
    return this._value.id;
  }

  get type(): ViewType {
    return this._value.type;
  }

  get title(): string {
    return this._value.title;
  }

  get summary(): string {
    return this._value.summary;
  }

  get description(): string {
    return this._value.description;
  }

  get search(): Search {
    return this._value.search;
  }

  get properties(): Properties {
    return this._value.properties;
  }

  get state(): ViewStateMap {
    return this._value.state;
  }

  get createdAt(): Date {
    return this._value.createdAt;
  }

  get widgetMapping(): WidgetMapping {
    return (this.state || Immutable.Map()).valueSeq().map((s) => s.widgetMapping).reduce((prev, cur) => Immutable.fromJS(prev).merge(Immutable.fromJS(cur)));
  }

  get owner(): string {
    return this._value.owner;
  }

  get requires(): Requirements {
    return this._value.requires || {};
  }

  getSearchTypeByWidgetId(widgetId: string): QuerySearchType | undefined | null {
    const widgetMapping = this.state.map((state) => state.widgetMapping).flatten(true);
    const searchTypeId = widgetMapping.get(widgetId).first();

    if (!searchTypeId) {
      throw new Error(`Search type for widget with id ${widgetId} does not exist`);
    }

    const searchTypes = flatten(this.search.queries.map((query) => query.searchTypes).toArray());

    return searchTypes.find((entry) => entry && entry.id && entry.id === searchTypeId);
  }

  getWidgetTitleByWidget(widget: Widget) {
    const widgetTitles = this.state.flatMap((state) => state.titles.get('widget'));
    const _defaultTitle = defaultTitle(widget);

    return widgetTitles.get(widget.id) || _defaultTitle;
  }

  toBuilder(): Builder {
    const { id, title, summary, description, search, properties, state, createdAt, owner, requires, type } = this._value;

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder(Immutable.Map({
      id,
      title,
      summary,
      description,
      search,
      properties,
      state,
      createdAt,
      owner,
      requires,
      type,
    }));
  }

  toJSON() {
    const { id, type, title, summary, description, search, properties, state, createdAt, owner } = this._value;

    return {
      id,
      type,
      title,
      summary,
      description,
      search_id: search.id,
      properties,
      state,
      created_at: createdAt,
      owner,
    } as unknown as ViewJson;
  }

  static fromJSON(value: ViewJson): View {
    const { id, type, title, summary, description, properties, state, created_at, owner, requires } = value;
    const viewState: ViewStateMap = Immutable.Map(state).map(ViewState.fromJSON).toMap();

    return View.create()
      .toBuilder()
      .id(id)
      .type(type)
      .title(title)
      .summary(summary)
      .description(description)
      .properties(properties)
      .state(viewState)
      .createdAt(created_at)
      .owner(owner)
      .requires(requires)
      .build();
  }

  static builder(): Builder {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder();
  }
}

type InternalBuilderState = Immutable.Map<string, any>;

class Builder {
  value: InternalBuilderState;

  constructor(value: InternalBuilderState = Immutable.Map()) {
    this.value = value;
  }

  id(value: string): Builder {
    return new Builder(this.value.set('id', value));
  }

  toNewView(): Builder {
    return new Builder(this.value.set('id', undefined).set('title', undefined));
  }

  type(value: ViewType): Builder {
    return new Builder(this.value.set('type', value));
  }

  newId(): Builder {
    return this.id(new ObjectID().toString());
  }

  title(value: string): Builder {
    return new Builder(this.value.set('title', value));
  }

  summary(value: string): Builder {
    return new Builder(this.value.set('summary', value));
  }

  description(value: string): Builder {
    return new Builder(this.value.set('description', value));
  }

  search(value: Search): Builder {
    return new Builder(this.value.set('search', value));
  }

  properties(value: Properties): Builder {
    return new Builder(this.value.set('properties', value));
  }

  state(value: (ViewStateMap | { [key: string]: ViewState })): Builder {
    return new Builder(this.value.set('state', Immutable.Map(value)));
  }

  createdAt(value: Date): Builder {
    return new Builder(this.value.set('createdAt', value));
  }

  owner(value: string): Builder {
    return new Builder(this.value.set('owner', value));
  }

  requires(value: Requirements): Builder {
    return new Builder(this.value.set('requires', value));
  }

  build(): View {
    const { id, type, title, summary, description, search, properties, state, createdAt, owner, requires } = this.value.toObject();

    return new View(id, type, title, summary, description, search, properties, state, createdAt, owner, requires);
  }
}
