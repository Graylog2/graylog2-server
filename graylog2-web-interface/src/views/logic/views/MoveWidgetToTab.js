// @flow strict
import * as Immutable from 'immutable';
import uuid from 'uuid/v4';

import type { QueryId } from 'views/logic/queries/Query';
import type { WidgetId } from 'views/logic/views/types';
import WidgetPosition from 'views/logic/widgets/WidgetPosition';
import type { TitlesMap } from 'views/stores/TitleTypes';

import View from './View';
import FindWidgetAndQueryIdInView from './FindWidgetAndQueryIdInView';
import UpdateSearchForWidgets from './UpdateSearchForWidgets';
import AddNewWidgetsToPositions from './AddNewWidgetsToPositions';

import Widget from '../widgets/Widget';

const _removeWidgetTitle = (titlesMap: TitlesMap, widgetId: WidgetId): TitlesMap => {
  const widgetTitles = titlesMap.get('widget');

  if (!widgetTitles) {
    return titlesMap;
  }

  const newWidgetTitles = widgetTitles.remove(widgetId);

  return titlesMap.set('widget', newWidgetTitles);
};

const _removeWidgetFromTab = (widgetId: WidgetId, queryId: QueryId, dashboard: View): View => {
  const viewState = dashboard.state.get(queryId);
  const widgetIndex = viewState.widgets.findIndex((widget) => widget.id === widgetId);
  const { widgetPositions, titles } = viewState;
  const newTitles = _removeWidgetTitle(titles, widgetId);

  delete widgetPositions[widgetId];
  const { widgetMapping } = viewState;
  const newWidgetMapping = widgetMapping.remove(widgetId);
  const newViewState = viewState.toBuilder()
    .widgets(viewState.widgets.delete(widgetIndex))
    .widgetMapping(newWidgetMapping)
    .titles(newTitles)
    .widgetPositions(widgetPositions)
    .build();

  return dashboard.toBuilder()
    .state(dashboard.state.set(queryId, newViewState))
    .build();
};

const _setWidgetTitle = (titlesMap: TitlesMap, widgetID: WidgetId, newTitle: ?string): TitlesMap => {
  if (!newTitle) {
    return titlesMap;
  }

  const widgetTitlesMap = titlesMap.get('widget', Immutable.Map());
  const newWidgetTitleMap = widgetTitlesMap.set(widgetID, newTitle);

  return titlesMap.set('widget', newWidgetTitleMap);
};

const _addWidgetToTab = (widget: Widget, targetQueryId: QueryId, dashboard: View, newWidgetPosition: WidgetPosition, widgetTitle: ?string): View => {
  const viewState = dashboard.state.get(targetQueryId);
  const newWidget = widget.toBuilder().id(uuid()).build();
  const newWidgets = viewState.widgets.push(newWidget);
  const overridePositions = Immutable.Map({ [newWidget.id]: newWidgetPosition });
  const { widgetPositions } = viewState;
  const newWidgetPositions = AddNewWidgetsToPositions(Immutable.Map(widgetPositions), newWidgets.toArray(), overridePositions);
  const newTitleMap = _setWidgetTitle(viewState.titles, newWidget.id, widgetTitle);
  const newViewState = viewState.toBuilder()
    .widgets(newWidgets)
    .titles(newTitleMap)
    .widgetPositions(newWidgetPositions)
    .build();

  return dashboard.toBuilder()
    .state(dashboard.state.set(targetQueryId, newViewState))
    .build();
};

const _getWidgetPosition = (widgetId: WidgetId, queryId: QueryId, view: View): WidgetPosition => {
  return view.state.get(queryId).widgetPositions[widgetId];
};

const _getWidgetTitle = (widgetId: WidgetId, queryId: QueryId, view: View): ?string => {
  return view.state.get(queryId).titles.get('widget').get(widgetId);
};

const MoveWidgetToTab = (widgetId: WidgetId, targetQueryId: QueryId, dashboard: View, copy: boolean = false): ?View => {
  if (dashboard.type !== View.Type.Dashboard) {
    throw new Error(`Unexpected type ${dashboard.type} expected ${View.Type.Dashboard}`);
  }

  const match: ?[Widget, QueryId] = FindWidgetAndQueryIdInView(widgetId, dashboard);

  if (match) {
    const [widget, queryId] = match;
    const widgetTitle = _getWidgetTitle(widgetId, queryId, dashboard);
    const newWidgetPosition = _getWidgetPosition(widgetId, queryId, dashboard).toBuilder()
      .col(1)
      .row(1)
      .build();
    const tempDashboard = copy ? dashboard : _removeWidgetFromTab(widgetId, queryId, dashboard);

    return UpdateSearchForWidgets(_addWidgetToTab(widget, targetQueryId, tempDashboard, newWidgetPosition, widgetTitle));
  }

  return undefined;
};

export default MoveWidgetToTab;
