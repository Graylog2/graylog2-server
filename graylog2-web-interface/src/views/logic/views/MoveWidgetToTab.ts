/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as Immutable from 'immutable';

import type { QueryId } from 'views/logic/queries/Query';
import type { WidgetId } from 'views/logic/views/types';
import type { TitlesMap } from 'views/stores/TitleTypes';
import type WidgetPosition from 'views/logic/widgets/WidgetPosition';

import View from './View';
import FindWidgetAndQueryIdInView from './FindWidgetAndQueryIdInView';
import UpdateSearchForWidgets from './UpdateSearchForWidgets';
import GenerateNextPosition from './GenerateNextPosition';

import type Widget from '../widgets/Widget';

const _removeWidgetTitle = (titlesMap: TitlesMap, widgetId: WidgetId): TitlesMap => {
  const widgetTitles = titlesMap.get('widget');

  if (!widgetTitles) {
    return titlesMap;
  }

  const newWidgetTitles = widgetTitles.remove(widgetId);

  return titlesMap.set('widget', newWidgetTitles);
};

const _removeWidgetFromTab = (widgetId: WidgetId, queryId: QueryId, dashboard: View): View => {
  const viewState = dashboard.state.get(queryId);
  const widgetIndex = viewState.widgets.findIndex((widget) => widget.id === widgetId);
  const { widgetPositions, titles } = viewState;
  const newTitles = _removeWidgetTitle(titles, widgetId);

  delete widgetPositions[widgetId];
  const { widgetMapping } = viewState;
  const newWidgetMapping = widgetMapping.remove(widgetId);
  const newViewState = viewState.toBuilder()
    .widgets(viewState.widgets.delete(widgetIndex))
    .widgetMapping(newWidgetMapping)
    .titles(newTitles)
    .widgetPositions(widgetPositions)
    .build();

  return dashboard.toBuilder()
    .state(dashboard.state.set(queryId, newViewState))
    .build();
};

const _setWidgetTitle = (titlesMap: TitlesMap, widgetID: WidgetId, newTitle: string | undefined | null): TitlesMap => {
  if (!newTitle) {
    return titlesMap;
  }

  const widgetTitlesMap = titlesMap.get('widget', Immutable.Map());
  const newWidgetTitleMap = widgetTitlesMap.set(widgetID, newTitle);

  return titlesMap.set('widget', newWidgetTitleMap);
};

const _addWidgetToTab = (widget: Widget, targetQueryId: QueryId, dashboard: View, widgetTitle: string | undefined | null, oldPosition: WidgetPosition): View => {
  const viewState = dashboard.state.get(targetQueryId);
  const newWidget = widget?.id ? widget : widget.toBuilder().newId().build();
  const newWidgets = viewState.widgets.push(newWidget);
  const { widgetPositions } = viewState;
  const widgetPositionsMap = oldPosition ? {
    ...widgetPositions,
    [newWidget.id]: oldPosition.toBuilder().row(0).col(0).build(),
  } : widgetPositions;
  const newWidgetPositions = GenerateNextPosition(Immutable.Map(widgetPositionsMap), newWidgets.toArray());
  const newTitleMap = _setWidgetTitle(viewState.titles, newWidget.id, widgetTitle);
  const newViewState = viewState.toBuilder()
    .widgets(newWidgets)
    .titles(newTitleMap)
    .widgetPositions(newWidgetPositions)
    .build();

  return dashboard.toBuilder()
    .state(dashboard.state.set(targetQueryId, newViewState))
    .build();
};

const _getWidgetTitle = (widgetId: WidgetId, queryId: QueryId, view: View): string | undefined | null => {
  return view.state.get(queryId).titles.getIn(['widget', widgetId]);
};

const MoveWidgetToTab = (widgetId: WidgetId, targetQueryId: QueryId, dashboard: View, copy: boolean = false): View | undefined | null => {
  if (dashboard.type !== View.Type.Dashboard) {
    throw new Error(`Unexpected type ${dashboard.type} expected ${View.Type.Dashboard}`);
  }

  const match: [Widget, QueryId] | undefined | null = FindWidgetAndQueryIdInView(widgetId, dashboard);

  if (match) {
    const [widget, queryId] = match;
    const widgetTitle = _getWidgetTitle(widgetId, queryId, dashboard);
    const { widgetPositions } = dashboard.state.get(queryId);
    const oldPosition = widgetPositions[widgetId];

    const tempDashboard = copy ? dashboard : _removeWidgetFromTab(widgetId, queryId, dashboard);
    const newWidget = copy ? widget.toBuilder().newId().build() : widget;

    return UpdateSearchForWidgets(_addWidgetToTab(newWidget, targetQueryId, tempDashboard, widgetTitle, oldPosition));
  }

  return undefined;
};

export default MoveWidgetToTab;
