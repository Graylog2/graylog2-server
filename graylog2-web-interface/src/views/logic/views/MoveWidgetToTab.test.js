import { readFileSync } from 'fs';

import { dirname } from 'path';

import MoveWidgetToTab from './MoveWidgetToTab';
import View from './View';

import Parameter from '../parameters/Parameter';
import ValueParameter from '../parameters/ValueParameter';
import Search from '../search/Search';

jest.mock('uuid/v4', () => jest.fn(() => 'dead-beef'));

jest.mock('bson-objectid', () => jest.fn(() => ({
  toString: jest.fn(() => 'new-search-id'),
})));

jest.mock('../Widgets', () => ({
  widgetDefinition: () => ({ searchTypes: () => [{ type: 'pivot' }] }),
}));

jest.mock('../SearchType', () => jest.fn(() => ({
  type: 'pivot',
  handler: jest.fn(),
  defaults: {},
})));

const cwd = dirname(__filename);
const readFixture = (filename) => JSON.parse(readFileSync(`${cwd}/${filename}`).toString());

describe('MoveWidgetToTab', () => {
  beforeEach(() => {
    Parameter.registerSubtype(ValueParameter.type, ValueParameter);
  });

  it('should move a Widget to a dashboard', () => {
    const dashboardFixture = View.fromJSON(readFixture('./MoveWidgetToTab.Dashboard.fixture.json'));
    const searchFixture = Search.fromJSON(readFixture('./MoveWidgetToTab.Search.fixture.json'));
    const dashboaad = dashboardFixture.toBuilder()
      .search(searchFixture)
      .build();

    const newDashboard = MoveWidgetToTab(
      'b34c3c6f-c49d-41d3-a65a-f746134f8f3e',
      '5faea09b-4187-4eda-9d59-7a86d4774c73',
      dashboaad,
      false,
    );

    expect(newDashboard).toMatchSnapshot();
  });

  it('should copy a Widget to a dashboard', () => {
    const dashboardFixture = View.fromJSON(readFixture('./MoveWidgetToTab.Dashboard.fixture.json'));
    const searchFixture = Search.fromJSON(readFixture('./MoveWidgetToTab.Search.fixture.json'));
    const dashboaad = dashboardFixture.toBuilder()
      .search(searchFixture)
      .build();

    const newDashboard = MoveWidgetToTab(
      'b34c3c6f-c49d-41d3-a65a-f746134f8f3e',
      '5faea09b-4187-4eda-9d59-7a86d4774c73',
      dashboaad,
      true,
    );

    expect(newDashboard).toMatchSnapshot();
  });
});
