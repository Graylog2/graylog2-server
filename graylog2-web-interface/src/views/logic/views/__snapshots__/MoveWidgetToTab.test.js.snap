// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoveWidgetToTab should copy a Widget to a dashboard 1`] = `
Object {
  "created_at": "2020-05-27T12:32:49.786Z",
  "description": "move widget to tab",
  "id": "5ece5e1b74ff709b0b0dc69b",
  "owner": "admin",
  "properties": Immutable.List [],
  "search_id": "new-search-id",
  "state": Immutable.Map {
    "fa247d8f-7afa-45b7-b57e-3cdef4ee53be": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {
        "b34c3c6f-c49d-41d3-a65a-f746134f8f3e": Object {
          "col": 1,
          "height": 4,
          "row": 1,
          "width": 4,
        },
      },
      "selected_fields": undefined,
      "titles": Immutable.Map {
        "widget": Immutable.Map {
          "b34c3c6f-c49d-41d3-a65a-f746134f8f3e": "Widget Title",
        },
      },
      "widget_mapping": Immutable.Map {
        "b34c3c6f-c49d-41d3-a65a-f746134f8f3e": Immutable.Set [
          "dead-beef",
        ],
      },
      "widgets": Immutable.List [
        Object {
          "config": Object {
            "column_pivots": Array [],
            "event_annotation": false,
            "rollup": true,
            "row_pivots": Array [],
            "series": Array [
              Object {
                "config": Object {
                  "name": "Message Count",
                },
                "function": "count()",
              },
            ],
            "sort": Array [],
            "visualization": "numeric",
          },
          "filter": undefined,
          "id": "b34c3c6f-c49d-41d3-a65a-f746134f8f3e",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
      ],
    },
    "5faea09b-4187-4eda-9d59-7a86d4774c73": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {
        "dead-beef": Object {
          "col": 0,
          "height": 4,
          "row": 1,
          "width": 4,
        },
      },
      "selected_fields": undefined,
      "titles": Immutable.Map {
        "widget": Immutable.Map {
          "dead-beef": "Widget Title",
        },
      },
      "widget_mapping": Immutable.Map {
        "dead-beef": Immutable.Set [
          "dead-beef",
        ],
      },
      "widgets": Immutable.List [
        Object {
          "config": Object {
            "column_pivots": Array [],
            "event_annotation": false,
            "rollup": true,
            "row_pivots": Array [],
            "series": Array [
              Object {
                "config": Object {
                  "name": "Message Count",
                },
                "function": "count()",
              },
            ],
            "sort": Array [],
            "visualization": "numeric",
          },
          "filter": undefined,
          "id": "dead-beef",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
      ],
    },
  },
  "summary": "move widget to tab",
  "title": "MoveWidgetToTab",
  "type": "DASHBOARD",
}
`;

exports[`MoveWidgetToTab should move a Widget to a dashboard 1`] = `
Object {
  "created_at": "2020-05-27T12:32:49.786Z",
  "description": "move widget to tab",
  "id": "5ece5e1b74ff709b0b0dc69b",
  "owner": "admin",
  "properties": Immutable.List [],
  "search_id": "new-search-id",
  "state": Immutable.Map {
    "fa247d8f-7afa-45b7-b57e-3cdef4ee53be": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {},
      "selected_fields": undefined,
      "titles": Immutable.Map {
        "widget": Immutable.Map {},
      },
      "widget_mapping": Immutable.Map {},
      "widgets": Immutable.List [],
    },
    "5faea09b-4187-4eda-9d59-7a86d4774c73": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {
        "dead-beef": Object {
          "col": 0,
          "height": 4,
          "row": 1,
          "width": 4,
        },
      },
      "selected_fields": undefined,
      "titles": Immutable.Map {
        "widget": Immutable.Map {
          "dead-beef": "Widget Title",
        },
      },
      "widget_mapping": Immutable.Map {
        "dead-beef": Immutable.Set [
          "dead-beef",
        ],
      },
      "widgets": Immutable.List [
        Object {
          "config": Object {
            "column_pivots": Array [],
            "event_annotation": false,
            "rollup": true,
            "row_pivots": Array [],
            "series": Array [
              Object {
                "config": Object {
                  "name": "Message Count",
                },
                "function": "count()",
              },
            ],
            "sort": Array [],
            "visualization": "numeric",
          },
          "filter": undefined,
          "id": "dead-beef",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
      ],
    },
  },
  "summary": "move widget to tab",
  "title": "MoveWidgetToTab",
  "type": "DASHBOARD",
}
`;

exports[`MoveWidgetToTab should provide a default position in case the widgets position cannot be found 1`] = `
Object {
  "created_at": "2020-05-27T12:32:49.786Z",
  "description": "move widget to tab",
  "id": "5ece5e1b74ff709b0b0dc69b",
  "owner": "admin",
  "properties": Immutable.List [],
  "search_id": "new-search-id",
  "state": Immutable.Map {
    "fa247d8f-7afa-45b7-b57e-3cdef4ee53be": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {},
      "selected_fields": undefined,
      "titles": Immutable.Map {
        "widget": Immutable.Map {},
      },
      "widget_mapping": Immutable.Map {},
      "widgets": Immutable.List [],
    },
    "5faea09b-4187-4eda-9d59-7a86d4774c73": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {
        "dead-beef": Object {
          "col": 1,
          "height": 4,
          "row": 1,
          "width": 4,
        },
      },
      "selected_fields": undefined,
      "titles": Immutable.Map {
        "widget": Immutable.Map {
          "dead-beef": "Widget Title",
        },
      },
      "widget_mapping": Immutable.Map {
        "dead-beef": Immutable.Set [
          "dead-beef",
        ],
      },
      "widgets": Immutable.List [
        Object {
          "config": Object {
            "column_pivots": Array [],
            "event_annotation": false,
            "rollup": true,
            "row_pivots": Array [],
            "series": Array [
              Object {
                "config": Object {
                  "name": "Message Count",
                },
                "function": "count()",
              },
            ],
            "sort": Array [],
            "visualization": "numeric",
          },
          "filter": undefined,
          "id": "dead-beef",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
      ],
    },
  },
  "summary": "move widget to tab",
  "title": "MoveWidgetToTab",
  "type": "DASHBOARD",
}
`;

exports[`MoveWidgetToTab should work when titles are empty 1`] = `
Object {
  "created_at": "2020-05-27T12:32:49.786Z",
  "description": "move widget to tab",
  "id": "5ece5e1b74ff709b0b0dc69b",
  "owner": "admin",
  "properties": Immutable.List [],
  "search_id": "new-search-id",
  "state": Immutable.Map {
    "fa247d8f-7afa-45b7-b57e-3cdef4ee53be": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {},
      "selected_fields": undefined,
      "titles": Immutable.Map {},
      "widget_mapping": Immutable.Map {},
      "widgets": Immutable.List [],
    },
    "5faea09b-4187-4eda-9d59-7a86d4774c73": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {
        "dead-beef": Object {
          "col": 0,
          "height": 4,
          "row": 1,
          "width": 4,
        },
      },
      "selected_fields": undefined,
      "titles": Immutable.Map {},
      "widget_mapping": Immutable.Map {
        "dead-beef": Immutable.Set [
          "dead-beef",
        ],
      },
      "widgets": Immutable.List [
        Object {
          "config": Object {
            "column_pivots": Array [],
            "event_annotation": false,
            "rollup": true,
            "row_pivots": Array [],
            "series": Array [
              Object {
                "config": Object {
                  "name": "Message Count",
                },
                "function": "count()",
              },
            ],
            "sort": Array [],
            "visualization": "numeric",
          },
          "filter": undefined,
          "id": "dead-beef",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
      ],
    },
  },
  "summary": "move widget to tab",
  "title": "MoveWidgetToTab",
  "type": "DASHBOARD",
}
`;
